lib/lm_docstache/parser.rb

```diff
@@ -0,0 +1,118 @@
+    BLOCK_OPERATOR_PATTERN = '\s*(~=|==)\s*'
+    BLOCK_VALUE_PATTERN = '([^\}]+?)\s*'
+    BLOCK_START_PATTERN = "{{#{BLOCK_TYPE_PATTERN}#{BLOCK_VARIABLE_PATTERN}"\
+                          "#{BLOCK_OPERATOR_PATTERN}#{BLOCK_VALUE_PATTERN}}}"
+    BLOCK_CONTENT_PATTERN = '(.*?)'
+    BLOCK_CLOSE_PATTERN = '{{/\s*\k<2>\s*}}'
+    BLOCK_NAMED_CLOSE_PATTERN = '{{/\s*%{tag_name}\s*}}'
+    BLOCK_PATTERN = "#{BLOCK_START_PATTERN}#{BLOCK_CONTENT_PATTERN}"\
+                    "#{BLOCK_CLOSE_PATTERN}"
+
+    BLOCK_START_MATCHER = /#{BLOCK_START_PATTERN}/
+    BLOCK_CLOSE_MATCHER = /{{\/\s*.+?\s*}}/
+    BLOCK_MATCHER = /#{BLOCK_PATTERN}/
+    VARIABLE_MATCHER = /{{([\w\.-]+)}}/
+
+    attr_reader :document, :data, :blocks
+
+    def initialize(document, data)
+      @document = document
+      @data = data.transform_keys(&:to_s)
+    end
+
+    def parse_and_update_document!
+      find_blocks
+      replace_conditional_blocks_in_document!
+      replace_variables_in_document!
+    end
+
+    private
+
+    def find_blocks
+      return @blocks if instance_variable_defined?(:@blocks)
+      return @blocks = [] unless document.text =~ BLOCK_MATCHER
+
+      @blocks = []
+      paragraphs = document.css('w|p')
+
+      while paragraph = paragraphs.shift do
+        content = paragraph.text
+        full_match = BLOCK_MATCHER.match(content)
+        start_match = !full_match && BLOCK_START_MATCHER.match(content)
+
+        next unless full_match || start_match
+
+        if full_match
+          @blocks.push(*ConditionalBlock.inline_blocks_from_paragraph(paragraph))
+        else
+          condition = condition_from_match_data(start_match)
+          comprised_paragraphs = all_block_elements(start_match[2], paragraph, paragraphs)
+
+          # We'll ignore conditional blocks that have no correspondent closing tag
+          next unless comprised_paragraphs
+
+          @blocks << ConditionalBlock.new(
+            elements: comprised_paragraphs,
+            condition: condition
+          )
+        end
+      end
+
+      @blocks
+    end
+
+    # Evaluates all conditional blocks inside the given XML document and keep or
+    # remove their content inside the document, depending on the truthiness of
+    # the condition on each given conditional block.
+    def replace_conditional_blocks_in_document!
+      blocks.each do |conditional_block|
+        value = data[conditional_block.condition.left_term]
+        conditional_block.evaluate_with_value!(value)
+      end
+    end
+
+    # It simply replaces all the referenced variables insithe document by their
+    # correspondent values provided in the attributes hash +data+.
+    def replace_variables_in_document!
+      document.css('w|t').each do |text_node|
+        text = text_node.text
+
+        next unless text =~ VARIABLE_MATCHER
+
+        text.gsub!(VARIABLE_MATCHER) { |_match| data[$1].to_s }
+        text_node.content = text
+      end
+    end
+
+    def condition_from_match_data(match)
```

Comment:
Over here is very welcome to also have examples.

Reply:
Done in 640e537.